### DISCUSSION TITLE
mx.viz.plot_network KeyError: 'ctc_loss0_output'

### DISCUSSION THREAD

**Discussion by MenglingD** [apache/incubator-mxnet#14548] [Q&A]
## Description

I builded my network with `mx.sym.ctc_loss` for the final loss on speech recognition task, and with `mx.viz.plot_network` to make sure the right struct of network. But I got an exception as I invoke `mx.viz.plot_network`, even if the symbol of network is as simple as possible.

## Environment info (Required)

operating system: ubuntu-16.04 LTS
python version: python 3.5
mxnet version: 1.4.0-cpu

## Error Message:

```
Traceback (most recent call last):
  File "ctc_plot.py", line 21, in <module>
    mx.viz.plot_network(sym, shape=shape).view()
  File "/home/windog/.local/lib/python3.5/site-packages/mxnet/visualization.py", line 382, in plot_network
    shape = shape_dict[key][1:]
KeyError: 'ctc_loss0_output'
```

## Minimum reproducible example
(If you are using your own code, please provide a short script that reproduces the error. Otherwise, please provide link to the existing example.)

**ctc_plot.py**:

``` python
import mxnet as mx

def ctc_sym():
    data = mx.sym.Variable('data')
    label = mx.sym.Variable('label')
    data_len = mx.sym.Variable('data_len')
    label_len = mx.sym.Variable('label_len')

    ctc_loss = mx.sym.ctc_loss(data, label, data_len, label_len, True, True)
    ctc_loss = mx.sym.MakeLoss(ctc_loss)

    return ctc_loss


if __name__ == '__main__':
    shape = {'data': (400, 32, 100),    # [max_seq_len, batch_size, num_cls]
             'label': (32, 50),         # [batch_size, max_label_len]
             'data_len': (32,),         # [batch_size,]
             'label_len': (32,)}        # [batch_size,]
    sym = ctc_sym()
    mx.viz.plot_network(sym, shape=shape).view()
```

## Steps to reproduce
(Paste the commands you ran that produced the error.)

1. python3 ctc_plot.py

## What have you tried to solve it?

1. there is a related issue: [#3221](https://github.com/apache/incubator-mxnet/issues/3221)
2. I have debugged into `incubator-mxnet/python/mxnet/visualization.py` in which throw a exception, and got a wrong `key` for `shape_dict` in following position in `visualization.py`:

![image](https://user-images.githubusercontent.com/9418558/55126339-b56f2a00-5147-11e9-9bcc-1b9d46482598.png)

And then, I found out that the input name of symbol of `MakeLoss(ctc_loss)`, `"ctc_loss0_output0"`, has a bit difference from `key = input_name + "_output"` which is `"ctc_loss0_output"`.  And shape variable `shape_dict` is given by following code in `visualization.py`:

![image](https://user-images.githubusercontent.com/9418558/55127050-ccaf1700-5149-11e9-962b-d3f90a63cebc.png)

The definition of `list_outputs` which is located at line 734 of `mxnet/model/symbol.py`:

![image](https://user-images.githubusercontent.com/9418558/55127012-b012df00-5149-11e9-878b-0ae90fe83b9a.png)

I could not deep into as its' low-level implement, so is there anyone to help me out. What does raise that exception? Incorrect invoking of `ctc_loss` or `plot_network`, or some inner mistakes in mxnet?

**Comment by MenglingD** [comment]
@szha

**Comment by vdantu** [comment]
@mxnet-label-bot add [question]

**Comment by szha** [comment]
@windoman sorry for the late reply. The index comes from that ctc_loss op is set to return more than one output: https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/ctc_loss.cc#L103

**Comment by MenglingD** [comment]
> @windoman sorry for the late reply. The index comes from that ctc_loss op is set to return more than one output: https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/ctc_loss.cc#L103

@szha Thank you very much for your reply. The `ctc_loss` have two outputs, a loss symbol for per sample in batch and gradient output, which are accounted [https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/ctc_loss.cc#L105](https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/ctc_loss.cc#L105), but only one output is visible [https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/ctc_loss.cc#L109](https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/ctc_loss.cc#L109).
